euphonic/sampling.py

```diff
@@ -0,0 +1,352 @@
+
+    The points form a grid in spherical coordinates
+
+    Parameters
+    ----------
+
+    n_phi
+        Number of longitude-like radial subdivisions at poles
+
+    n_theta
+        Number of lattitude-like rings of points dividing polar axis
+
+    cartesian
+        Yield points in Cartesian coordinates. If False, instead yield points
+        in spherical coordinates.
+
+    jitter
+        Randomly displace each point within its own "cell" of the grid
+
+    Returns
+    -------
+    Iterator[Tuple[float, float, float]]
+
+        Sequence of (x, y, z) coordinates (if cartesian=True) or
+        (r, phi, theta) spherical coordinates.
+    """
+    phi_sequence = np.linspace(-np.pi, np.pi, n_phi + 1)[:-1]
+    phi_spacing = phi_sequence[1] - phi_sequence[0]
+
+    # Offset theta samples to avoid redundancy at pole
+    theta_sequence = (np.linspace(0, np.pi, n_theta + 1)[:-1]
+                      + np.pi / (2 * n_theta))
+    theta_spacing = theta_sequence[1] - theta_sequence[0]
+
+    for phi, theta in product(phi_sequence, theta_sequence):
+        if jitter:
+            displacement = ((np.random.random(2) - 0.5)
+                            * [phi_spacing, theta_spacing])
+            phi, theta = [phi, theta] + displacement
+
+        if cartesian:
+            yield _spherical_polar_to_cartesian(phi, theta)
+        else:
+            yield (1, phi, theta)
+
+
+def spherical_polar_improved(npts: int,
+                             cartesian: bool = True, jitter: bool = False,
+                             ) -> Iterator[Tuple[float, float, float]]:
+    """Yield a series of 3D points on a unit sphere surface
+
+    The points form rings of common theta in polar coordinates. However, the
+    number of samples in each ring is scaled to achieve as uniform a sampling
+    density as possible.
+
+    Analytically we find that for evenly-divided theta, the average length of a
+    "latitude" line is 4.
+
+    To obtain parity between nearest-neighbours (NN) along each axis,
+    we solve (NN distance along theta) = (NN distance along phi):
+
+    4 n_theta / npts = pi / n_theta
+
+    Exact solutions with integer npts will be an irrational number, so n_theta
+    is rounded down. The requested npts is then distributed between the
+    constant-theta rings, according to their circumference.
+
+    Parameters
+    ----------
+
+    npts
+        Number of points at sphere surface
+
+    cartesian
+        Yield points in Cartesian coordinates. If False, instead yield points
+        in spherical coordinates.
+
+    jitter 
+        Randomly displace each point within its own "cell" of the irregular
+        grid
+
+    Returns
+    -------
+    Iterator[Tuple[float, float, float]]
+
+        Sequence of (x, y, z) coordinates (if cartesian=True) or
+        (r, phi, theta) spherical coordinates.
+
+    """
+    if npts < 6:
+        raise ValueError("This sampling scheme has a minimum of 6 points")
```

Comment:
Add `Raises` section to docstring

Reply:
Good spot. Should I also add it to the powder methods that call this function?

Reply:
That's a good question. I would say probably not as it should be fairly uncommon. Also in this case if we add raises documentation further up the stack that might set a precedent for following other exceptions through elsewhere...